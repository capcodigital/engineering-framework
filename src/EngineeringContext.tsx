import { createContext, useState, FC } from "react";
export type EngineeringContextType = {
  specialism: string;
  level: string;
  category: string;
  competency: string;
  setSpecialism: (newSpecialism: string) => void;
  setLevel: (newLevel: string) => void;
  setCategory: (newCategory: string) => void;
  setCompetency: (newComptenecy: string) => void;
};

export const EngineeringContext = createContext<EngineeringContextType>({
  specialism: "",
  level: "",
  category: "",
  competency: "",
  setSpecialism: () => {},
  setLevel: () => {},
  setCategory: () => {},
  setCompetency: () => {},
});

export const EngineeringContextProvider: FC = ({ children }) => {
  const [specialism, setSpecialism] = useState("Software Engineer");
  const [level, setLevel] = useState("Associate");
  const [category, setCategory] = useState("Overview");
  const [competency, setCompetency] = useState("Description");
  const initialValue = {
    specialism,
    level,
    category,
    competency,
    setSpecialism,
    setLevel,
    setCategory,
    setCompetency,
  };

  return (
    <EngineeringContext.Provider value={initialValue}>
      {children}
    </EngineeringContext.Provider>
  );
};
