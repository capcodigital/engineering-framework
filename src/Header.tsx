import React, { useState, useContext } from 'react';
import { Link } from 'react-router-dom';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import styled from 'styled-components/macro';
import {EngineeringContext} from './EngineeringContext';
import logo from './logo.png';
import Drawer from '@material-ui/core/Drawer';
import List from '@material-ui/core/List';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import ClearIcon from '@material-ui/icons/Clear';
import { ListItem } from '@material-ui/core';

const ListWrapper = styled.div`
  width: 200px;
  margin: 20px;
  font-size: 20px;
`;

const StyledAppBar = styled(AppBar)`
  && {
    background: black;
    padding-top: 20px;
  }
`;

const StyledIconButton = styled(IconButton)`
  && {
    @media screen and (min-width: 600px) {
      display: none;
    }
  }
`;

const StyledLink = styled(Link)`
  font-size: 16px;
  font-weight: bold;
  position: relative;
  text-decoration: none;
  color: grey;
  margin-left: 2rem;
  &.active {
    color: white;
  }
  :hover {
    color: lightgrey;
  }
  @media screen and (max-width: 600px) {
    display: none;
  }
`;

const titleData = [
  { name: 'Software Engineer', url: '/software' },
  { name: 'Quality Engineer', url: '/quality' },
];

export default function Header() {
  const { specialism, setSpecialism } = useContext(EngineeringContext);
  const [open, setOpen] = useState(false);

  const handleClick = (activeTitle: string) => {
    setSpecialism(activeTitle);
  };

  const toggleDrawer = (open: boolean) => (event: any) => {
    if (
      event.type === 'keydown' &&
      (event.key === 'Tab' || event.key === 'Shift')
    ) {
      return;
    }
    setOpen(open);
  };

  return (
    <>
      <StyledAppBar position='static'>
        <Toolbar>
          <StyledIconButton
            onClick={toggleDrawer(true)}
            edge='start'
            color='inherit'
            aria-label='menu'
          >
            <MenuIcon fontSize='large' />
          </StyledIconButton>
          <img height={30} src={logo} alt='Logo' />
          {titleData.map((link: any) =>
            link.name === specialism ? (
              <StyledLink
                key={link.name}
                to={link.url}
                className={'active'}
                onClick={() => handleClick(link.name)}
              >
                {link.name}
              </StyledLink>
            ) : (
              <StyledLink
                key={link.name}
                to={link.url}
                onClick={() => handleClick(link.name)}
              >
                {link.name}
              </StyledLink>
            )
          )}
        </Toolbar>
      </StyledAppBar>
      <Drawer anchor={'left'} open={open} onClose={toggleDrawer(false)}>
        <ListWrapper
          role='presentation'
          onClick={toggleDrawer(false)}
          onKeyDown={toggleDrawer(false)}
        >
          <IconButton
            onClick={toggleDrawer(false)}
            edge='end'
            color='inherit'
            aria-label='clear'
          >
            <ClearIcon fontSize='large' />
          </IconButton>
          <List>
            {titleData.map((link: any) => (
              <ListItem
                key={link.name}
                button
                component={Link}
                to={link.url}
                onClick={() => handleClick(link.name)}
              >
                {link.name}
              </ListItem>
            ))}
          </List>
        </ListWrapper>
      </Drawer>
    </>
  );
}
