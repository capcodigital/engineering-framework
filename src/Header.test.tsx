import * as React from 'react';
import { render, screen } from '@testing-library/react';
import Header from './Header';
import { BrowserRouter as Router } from 'react-router-dom';
import { EngineeringContextProvider } from './EngineeringContext';
it('should render Header component', () => {
  const { container } = render(
    <Router>
      <Header />
    </Router>
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('should check that software engineer menu exists', () => {
  render(
    <Router>
      <Header />
    </Router>
  );
  expect(screen.getByText('Software Engineer')).toBeTruthy();
});

it('should check that quality engineer menu exists', () => {
  render(
    <Router>
      <Header />
    </Router>
  );
  expect(screen.getByText('Quality Engineer')).toBeTruthy();
});

it('should check that Software Engineer is the active menu option', () => {
 render(
    <Router>
      <EngineeringContextProvider>
        <Header />
      </EngineeringContextProvider>
    </Router>
  );
  expect(screen.getByText('Software Engineer').classList).toContain('active');
});
