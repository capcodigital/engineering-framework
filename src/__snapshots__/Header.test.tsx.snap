// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render Header component 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary Header__StyledAppBar-u4z9pp-0 dEsAkp MuiPaper-elevation4"
>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <button
      aria-label="menu"
      class="MuiButtonBase-root MuiIconButton-root Header__StyledIconButton-u4z9pp-1 iRuLqq MuiIconButton-colorInherit MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <img
      alt="Logo"
      height="32"
      src="logo.png"
    />
    <a
      class="Header__StyledLink-u4z9pp-2 cMYGaa"
      href="/software/#-"
    >
      Software Engineer
    </a>
    <a
      class="Header__StyledLink-u4z9pp-2 cMYGaa"
      href="/quality/#-"
    >
      Quality Engineer
    </a>
  </div>
</header>
`;
